9armbot/9armbot.js

```diff
@@ -172,6 +217,73 @@ function gacha(channel, user, amount) {
         webapp.socket.io().emit("widget::killfeed", {
             message: killfeed_msg,
         });
+
+        const txnPayload = {
+            username: _player.username,
+            amount: amount,
+            gain: gain || 0,
+            rate: gachaRateType,
+            timestamp: dayjs().unix(),
+            txnTime: dayjs().format("D MMM YYYY HH:mm:ss")
+        }
+
+        webapp.socket.io().emit("widget::market_dashboard", {
+            key: MARKET_KEY.TRANSACTION,
+            data: txnPayload
+        });
+
+        if(previousSessionIncome !== sessionIncome){
+            pushSessionData(sessionIncomeDataSet, {
+                t: txnPayload.timestamp,
+                y: sessionIncome
+            })
+            webapp.socket.io().emit("widget::market_dashboard", {
+                key: MARKET_KEY.SESSION_INCOME,
+                data: {
+                    timestamp: txnPayload.timestamp,
+                    txnTime: txnPayload.txnTime,
+                    income: sessionIncome
+                }
+            });
+            previousSessionIncome = sessionIncome
+        }
+
+        if(previousSessionPayout !== sessionPayout){
+            pushSessionData(sessionPayoutDataSet, {
+                t: txnPayload.timestamp,
+                y: sessionPayout
+            })
+            webapp.socket.io().emit("widget::market_dashboard", {
+                key: MARKET_KEY.SESSION_PAYOUT,
+                data: {
+                    timestamp: txnPayload.timestamp,
+                    txnTime: txnPayload.txnTime,
+                    payout: sessionPayout
+                }
+            });
+            previousSessionPayout = sessionPayout
+        }
+
+
+        if(gachaWinners.length > 0){
+            webapp.socket.io().emit("widget::market_dashboard", {
+                key: MARKET_KEY.LATEST_WINNERS,
+                data: {
+                    winners: gachaWinners
+                }
+            });
+        }
+        
+        const currentTopCoinPlayers = player.getCoinTop(MIN_TOP_COIN_PLAYER)
+        if(currentTopCoinPlayers.length > 0 && gachaRateType !== GACHA_RATE_TYPE.SALT){
```

Comment:
`getCoinTop` function is undefined, you forgot add this in `player.js` ?

Reply:
Oops, wrong line comment.

Reply:
Hmm? 
`player` refers to variable at line 27 in `9armbot.js` that is used  many places in the file. 

If everything is ok, Please resolve this conversation.

Reply:
Sorry about this, I forgot fully check in player class

Reply:
No problem